VisualPinball.Unity/VisualPinball.Unity.Editor/Layers/LayerHandler.cs

```diff
@@ -1,36 +1,102 @@
 {
 	/// <summary>
-	/// This handler will construct a layer structure from the table loaded data.
-	/// It will then be in charge of layers management (add/remove/item assignation)
-	/// It will use the SceneVisibilityManager to synchronize layers/items visibility status & editor visibility
+	/// This handler will construct a layer structure from the table loaded data and store it into a TreeModel.
+	/// It will then be in charge of layers management (add/remove/rename/item assignation) by listening the TreeModel changed event
+	/// When TrereModel has changed it will reupdate all BiffData related to layers
 	/// </summary>
 	class LayerHandler
 	{
-		public bool IsVisible(GameObject gameObj)
+		/// <summary>
+		/// TreeModel used by the LayerTreeView, will be built based on the Layers structure
+		/// </summary>
+		public TreeModel<LayerTreeElement> TreeModel { get; private set; } = null;
+
+		/// <summary>
+		/// Readable version of the layers structure, will be used to build the TreeModel
+		/// </summary>
+		public Dictionary<string, List<Behaviour>> Layers { get; private set; } = new Dictionary<string, List<Behaviour>>();
+
+		public VisualPinball.Engine.VPT.Table.Table Table { get; private set; } = null;
+		
+		public LayerHandler()
 		{
-			return !SceneVisibilityManager.instance.IsHidden(gameObj);
+			//Initializing the TreeModel with Root item only until a table data is set
+			List<LayerTreeElement> elementList = new List<LayerTreeElement>();
+			elementList.Add(new LayerTreeElement() { Depth = -1, Id = -1 });
+			TreeModel = new TreeModel<LayerTreeElement>(elementList);
 		}
 
-		public void SetVisibility(GameObject gameObj, bool visibility, bool includeDescendants)
+		/// <summary>
+		/// Is called by the LayerEditor when a new TableBehavior is created/deleted
+		/// </summary>
+		/// <param name="tableBehavior"></param>
+		public void RebuildLayers(TableBehavior tableBehavior)
+		{
+			Table = tableBehavior != null ? tableBehavior.Table : null;
+			Layers.Clear();
+			if (tableBehavior != null) {
+				BuildLayersRecursively(tableBehavior.gameObject);
+			}
+
+			RebuildTreeModel();
+		}
+		private void BuildLayersRecursively(GameObject gameObj)
 		{
-			if (visibility) {
-				SceneVisibilityManager.instance.Show(gameObj, includeDescendants);
-			} else {
-				SceneVisibilityManager.instance.Hide(gameObj, includeDescendants);
+			for (int i = 0; i < gameObj.transform.childCount; ++i) {
+				var child = gameObj.transform.GetChild(i).gameObject;
+				foreach (var bh in child.GetComponents<Behaviour>()) {
+					AddToLayer(bh);
+				}
+				BuildLayersRecursively(child);
 			}
 		}
-		public void SetVisibility(GameObject[] gameObjs, bool visibility, bool includeDescendants)
+
+		private void AddToLayer(Behaviour bh)
+		{
+			if (bh is VPT.ILayerableItemBehavior layerableItemBehavior) {
+				if (layerableItemBehavior.EditorLayerName == "") {
+					layerableItemBehavior.EditorLayerName = $"Layer_{layerableItemBehavior.EditorLayer + 1}";

```

Comment:
Will this update the data as well? I.e. will the layer name be written to the `.vpx` when exporting?

Reply:
yup it will

Reply:
Cool!